pennylane_lightning_gpu/src/simulator/StateVectorCudaMPI.hpp

```diff
@@ -941,6 +955,364 @@ class StateVectorCudaMPI
+                    /* int64_t */ num_cols_local,
+                    /* int64_t */ nnz_local,
+                    /* void* */ d_csrOffsets.getData(),
+                    /* void* */ d_columns.getData(),
+                    /* void* */ d_values.getData(),
+                    /* cusparseIndexType_t */ compute_type,
+                    /* cusparseIndexType_t */ compute_type,
+                    /* cusparseIndexBase_t */ CUSPARSE_INDEX_BASE_ZERO,
+                    /* cudaDataType */ data_type));
+
+                // Create dense vector X
+                PL_CUSPARSE_IS_SUCCESS(cusparseCreateDnVec(
+                    /* cusparseDnVecDescr_t* */ &vecX,
+                    /* int64_t */ num_cols_local,
+                    /* void* */ BaseType::getData(),
+                    /* cudaDataType */ data_type));
+
+                // Create dense vector y
+                PL_CUSPARSE_IS_SUCCESS(cusparseCreateDnVec(
+                    /* cusparseDnVecDescr_t* */ &vecY,
+                    /* int64_t */ num_rows_local,
+                    /* void* */ d_tmp.getData(),
+                    /* cudaDataType */ data_type));
+
+                // allocate an external buffer if needed
+                PL_CUSPARSE_IS_SUCCESS(cusparseSpMV_bufferSize(
+                    /* cusparseHandle_t */ handle,
+                    /* cusparseOperation_t */ CUSPARSE_OPERATION_NON_TRANSPOSE,
+                    /* const void* */ &alpha,
+                    /* cusparseSpMatDescr_t */ mat,
+                    /* cusparseDnVecDescr_t */ vecX,
+                    /* const void* */ &beta,
+                    /* cusparseDnVecDescr_t */ vecY,
+                    /* cudaDataType */ data_type,
+                    /* cusparseSpMVAlg_t */ CUSPARSE_SPMV_ALG_DEFAULT,
+                    /* size_t* */ &bufferSize));
+
+                DataBuffer<cudaDataType_t, int> dBuffer{bufferSize, device_id,
+                                                        stream_id, true};
+
+                // execute SpMV
+                PL_CUSPARSE_IS_SUCCESS(cusparseSpMV(
+                    /* cusparseHandle_t */ handle,
+                    /* cusparseOperation_t */ CUSPARSE_OPERATION_NON_TRANSPOSE,
+                    /* const void* */ &alpha,
+                    /* cusparseSpMatDescr_t */ mat,
+                    /* cusparseDnVecDescr_t */ vecX,
+                    /* const void* */ &beta,
+                    /* cusparseDnVecDescr_t */ vecY,
+                    /* cudaDataType */ data_type,
+                    /* cusparseSpMVAlg_t */ CUSPARSE_SPMV_ALG_DEFAULT,
+                    /* void* */ reinterpret_cast<void *>(dBuffer.getData())));
+
+                // destroy matrix/vector descriptors
+                PL_CUSPARSE_IS_SUCCESS(cusparseDestroySpMat(mat));
+                PL_CUSPARSE_IS_SUCCESS(cusparseDestroyDnVec(vecX));
+                PL_CUSPARSE_IS_SUCCESS(cusparseDestroyDnVec(vecY));
+
+                color = 1;
```

Comment:
What does `color` do?

Reply:
It's used to categorize the `MPI_COMM_WORLD` into two groups. `color = 0` means  that `nnz` of current MPI rank is 0 and `color=1` means that `nnz` of current MPI rank is not zero. We just need to do `reduce` operation for the `color=1` group.